src/mlpack/methods/reinforcement_learning/environment/maze.hpp

```diff
@@ -0,0 +1,331 @@
+   *    terminates. If the value is 0, there is no limit.
+   * @param length Length of the binary vector for state and goal
+   */
+  Maze(const size_t maxSteps = 100) :
+      maxSteps(maxSteps),
+      stepsPerformed(0)
+  { 
+    maze = { { 0, -1, 0 ,1 },
+             { 0, -1, 0, 0 },
+             { 0, 0, 0, 0 }, 
+             { -1, -1, 0, -1 },};
+    
+    // for(size_t row=0 ; row < maze.n_rows ; ++row)
+    // {
+    //     for(size_t column = 0; column < maze.n_cols; ++column)
+    //     {
+    //         if(maze(row,column)==0){
+    //             startingPoints.push_back(arma::vec({row,column}));
+    //         }
+    //     }
+    // }
+    startingPoints.push_back(arma::vec({0,0}));
+    startingPoints.push_back(arma::vec({2,0}));
+    startingPoints.push_back(arma::vec({0,1}));
+    startingPoints.push_back(arma::vec({2,1}));
+    startingPoints.push_back(arma::vec({3,1}));
+    startingPoints.push_back(arma::vec({0,2}));
+    startingPoints.push_back(arma::vec({1,2}));
+    startingPoints.push_back(arma::vec({2,2}));
+    startingPoints.push_back(arma::vec({3,2}));
+    startingPoints.push_back(arma::vec({2,3}));
+
+  }
+
+  /**
+   * Get reward and next state based on current
+   * state and current action.
+   *
+   * @param state The current state.
+   * @param action The current action.
+   * @param nextState The next state.
+   * @return reward, it's always 1.0.
+   */
+  double Sample(const State& state,
+                const Action& action,
+                State& nextState,
+                const State& transitionGoal)
+  {
+    // Update the number of steps performed.
+    stepsPerformed++;
+
+    // Make a vector to estimate nextstate.
+    arma::vec currentState {state.Row(), state.Column()};
+    arma::vec direction = std::unordered_map<Action::actions, arma::vec>({
+            { Action::actions::left,  arma::vec({0, -1}) },
+            { Action::actions::right,  arma::vec({0, 1}) },
+            { Action::actions::up,  arma::vec({-1, 0}) },
+            { Action::actions::down,  arma::vec({1, 0}) }
+        })[action.action];
+
+    arma::vec currentNextState = currentState + direction;
+    nextState.Row() = currentNextState[0];
+    nextState.Column() = currentNextState[1];
+
+    // dont move to that position if invalid
+    bool invalid = false;
+
+    if(currentNextState(0)>= maze.n_rows || currentNextState(1)>= maze.n_cols || currentNextState(0) < 0 || currentNextState(1) < 0 || maze(currentNextState(0),currentNextState(1))==-1 ){
+        invalid = true;
+        nextState.Row() = currentState[0];
+        nextState.Column() = currentState[1];
+    }   
+
+    // Check if the episode has terminated.
+    bool done = IsTerminal(nextState);
+
+    // Do not reward agent if it failed.
+    if (done && maxSteps != 0 && stepsPerformed >= maxSteps)
+      return 0.0;
+    
+    if (done && maze(nextState.Row(),nextState.Column())==1 && nextState.Row()== goal(0) && nextState.Column()== goal(1) )
```

Comment:
Revisit our design guidelines

Reply:
Apologies ! I will make the style fixes ! i pushed the code so that you could have chance to look at it first hand.